kvbc/include/categorization/updates.h

```diff
@@ -0,0 +1,165 @@
+// Concord
+//
+// Copyright (c) 2020 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the
+// "License").  You may not use this product except in compliance with the
+// Apache 2.0 License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+#pragma once
+
+#include "assertUtils.hpp"
+#include "kv_types.hpp"
+#include "categorization/cmf/updates_cmf.hpp"
+
+#include <ctime>
+#include <map>
+#include <optional>
+#include <set>
+#include <string>
+#include <variant>
+
+// Categorized key-value updates for KVBC blocks. Every category supports different properties and functionalities.
+// Note1: Empty category IDs are invalid and not supported.
+// Note2: Using the same category ID for different category types is an error.
+namespace concord::kvbc::categorization {
+
+// Shared updates across multiple categories.
+// Persists key-values directly in the underlying key-value store. All key-values are marked stale during the update
+// itself. Explicit deletes are not supported.
+struct SharedKeyValueUpdates {
+  SharedKeyValueUpdates() { data_.stale_on_update = true; }
+  SharedKeyValueUpdates(SharedKeyValueUpdates&& other) {
+    data_.kv.swap(other.data_.kv);
+    data_.calculate_hash = other.data_.calculate_hash;
+    data_.stale_on_update = other.data_.stale_on_update;
+  }
+
+  struct SharedValue {
+    SharedValue(std::string&& val, std::set<std::string>&& cat_ids) {
+      data_.value = std::move(val);
+      for (auto it = cat_ids.begin(); it != cat_ids.end();) {
+        data_.categories_ids.emplace_back(cat_ids.extract(it++).value());
+      }
+    }
+
+   private:
+    SharedValueData data_;
+    friend class SharedKeyValueUpdates;
+  };
+
+  void addUpdate(std::string&& key, SharedValue&& val) { data_.kv.emplace(std::move(key), std::move(val.data_)); }
+
+ private:
+  SharedUpdatesData data_;
+};
+
+// Updates for a key-value category.
+// Persists key-values directly in the underlying key-value store.
+// Controls whether a hash of the updated key-values is calculated for this update
+struct KeyValueUpdates {
+  KeyValueUpdates() {}
+
+  KeyValueUpdates(KeyValueUpdates&& other) {
+    data_.kv.swap(other.data_.kv);
+    data_.deletes.swap(other.data_.deletes);
+    data_.calculate_hash = other.data_.calculate_hash;
+    unique_deletes_.swap(other.unique_deletes_);
+  }
+
+  struct ValueData {
```

Comment:
Nit: It came to mind we might call this `struct Value` and the member `std::string data`.